Template-based rendering system for item and mob descriptions

MorquinDevlar opened the pull request:
## Summary

This PR introduces a template-based rendering system for item and mob descriptions, replacing hardcoded string generation with template files. The system includes configurable equipment display options and error handling for edge cases.

## Changes

### Added

  - Template-based rendering system for items and mobs
    - ItemLookData struct with GetLookData() method for building template data
    - MobLookData struct with GetLookData() method for building template data
    - look-item.template and look-item.screenreader.template for item descriptions
    - look-mob.template and look-mob.screenreader.template for mob descriptions
    - Screenreader variants for character description and inventory-look templates
  - Configurable equipment display system
    - UserInterface config section with Formats and Display subsections
    - ShowEmptyEquipmentSlots config option to control equipment visibility
    - ShouldShowEmptyEquipmentSlots() config method
    - Template function showEmptyEquipmentSlots to access config from templates
    - "xxx is not wearing anything" message when mob has no equipment and showEmpty is false
  - Defensive programming
    - Nil checks in Item.GetLookData() to handle nil items
    - Nil checks in Mob.GetLookData() to handle nil mobs
    - Division-by-zero protection for mob health percentage calculations

### Changed

  - Look command implementation to use template rendering instead of hardcoded strings
  - Moved TextFormats config section to UserInterface.Formats
  - Default config indentation to use spaces (Every editor I've used wanted to do this automatically with every save - I gave in. Easy to change if it causes issues.)
  - MobLookData Equipment field to pointer type for template method compatibility

### Fixed

  - Template conditionals to handle all equipment display scenarios
  - Circular import by passing charmer name as parameter to GetLookData

### Removed

  - Hardcoded item and mob description generation in look command

### Configuration

  The ShowEmptyEquipmentSlots option:
  - true (default): Shows all equipment slots including empty ones
  - false: Only shows equipped items

### Breaking Changes

  - TextFormats config section moved to UserInterface.Formats - users with custom config files must update this path manually

MorquinDevlar committed:
Added template-based rendering for items and mobs

Added:
  - ItemLookData struct with GetLookData() method for template data
building
  - MobLookData struct with GetLookData() method for template data
building
  - look-item.template for rendering item descriptions
  - look-mob.template for rendering mob descriptions
  - Screenreader variants for all new templates
  - Character template screenreader variants for description and
inventory-look
  - Conditional equipment display in mob templates (only shows equipped
items)

Changed:
  - Look command to use templates for items and mobs instead of
hardcoded strings
  - MobLookData Equipment field to pointer type for template method
compatibility
  - Mob templates to only display equipment section when items are
equipped

Fixed:
  - Circular import issue by passing charmer name as parameter
  - Container reference removed from items package to avoid dependency
issues

MorquinDevlar committed:
Added configurable equipment display for mob template

Added:
  - UserInterface config section with Formats and Display subsections
  - ShowEmptyEquipmentSlots config option to control equipment display
  - Config method ShouldShowEmptyEquipmentSlots() for checking the
setting
  - Template function showEmptyEquipmentSlots to access config from
templates
  - GetUserInterfaceConfig() helper function
  - "xxx is not wearing anything" message when mob has no equipment and
showEmpty is false

Changed:
  - Moved TextFormats config under UserInterface.Formats for better
organization
  - Updated mob templates to respect ShowEmptyEquipmentSlots
configuration
  - Equipment display now shows all slots when true (default), only
equipped items when false

Fixed:
  - Template conditionals to properly handle all equipment display
scenarios
  - Both regular and screenreader templates now correctly show equipment
based on config

MorquinDevlar committed:
Refactored templates and improved defensive programming

Added:
  - Nil checks in Item.GetLookData() to handle nil items
  - Nil checks in Mob.GetLookData() to handle nil mobs
  - Division-by-zero protection for mob health calculations

Changed:
  - Consolidated equipment display logic in mob templates from 3 blocks
to 1
  - Used template variables to control equipment slot visibility
  - Reduced code duplication in both standard and screenreader templates

Removed:
  - TextFormats to UserInterface.Formats migration logic
  - Redundant equipment display blocks in mob templates

Breaking Changes:
  - Users with custom config files must manually update TextFormats to
UserInterface.Formats structure

MorquinDevlar committed:
Setting correct default settings in config.yaml

Volte6 reviewed (commented):
Some comments/questions/discussion - I haven't tested it yet, just going through the code.

MorquinDevlar commented on internal/items/items.go:
```diff
@@ -168,6 +168,76 @@ func (i *Item) Validate() {
+		}
+	}
+
+	iSpec := i.GetSpec()
+
+	data := ItemLookData{
+		Item:          i,
+		ItemSpec:      iSpec,
+		Location:      location,
+		IsWeapon:      iSpec.Type == Weapon,
+		IsReadable:    iSpec.Type == Readable,
+		IsDrinkable:   iSpec.Subtype == Drinkable,
+		IsEdible:      iSpec.Subtype == Edible,
+		IsLockpicks:   iSpec.Type == Lockpicks,
+		IsKey:         iSpec.Type == Key,
+		IsWearable:    iSpec.Subtype == Wearable,
+		IsUsable:      iSpec.Subtype == Usable,
+		HasUses:       iSpec.Uses > 0,
+		UsesRemaining: i.Uses,
+		MaxUses:       iSpec.Uses,
+		IsCursed:      i.IsCursed(),
+		IsEnchanted:   i.Enchantments > 0,
+		EnchantLevel:  int(i.Enchantments),
+	}
+
+	if data.IsWeapon {
+		data.WeaponHands = iSpec.Hands
+		data.WaitRounds = iSpec.WaitRounds
+
+		switch iSpec.Subtype {
+		case Claws:
+			data.WeaponType = "claws"
+		case Shooting:
+			data.WeaponType = "shooting"
+		}
+	}
+
+	return data
+}
+
```
Very good point. I'll change it to string comparison and add the adjective setup as well.

MorquinDevlar commented on internal/mobs/mobs.go:
```diff
@@ -708,6 +708,72 @@ func (m *Mob) GetScriptPath() string {
+	data.ItemCount = len(m.Character.Items)
+	if data.ItemCount == 0 {
+		data.CarryingStatus = "no"
+	} else if data.ItemCount < 4 {
+		data.CarryingStatus = "a few"
+	} else if data.ItemCount < 7 {
```
Overkill - I'll leave as is for now. It just adds extra code without real benefit.

MorquinDevlar commented on internal/mobs/mobs.go:
```diff
@@ -708,6 +708,72 @@ func (m *Mob) GetScriptPath() string {
+		data.HealthStatus = m.Character.Name + " appears to be in an unknown state."
+	} else {
+		healthPct := int(math.Ceil((float64(m.Character.Health) / float64(m.Character.HealthMax.Value)) * 100))
+		if healthPct >= 100 {
+			data.HealthStatus = m.Character.Name + " is in perfect health."
+		} else if healthPct >= 80 {
+			data.HealthStatus = m.Character.Name + " has a few scratches."
+		} else if healthPct >= 50 {
+			data.HealthStatus = m.Character.Name + " has some cuts and bruises."
+		} else if healthPct >= 15 {
```
Overkill - I'll leave as is for now. It just adds extra code without real benefit.

MorquinDevlar commented on internal/mobs/mobs.go:
```diff
@@ -708,6 +708,72 @@ func (m *Mob) GetScriptPath() string {
+	}
+
+	// Calculate health status with defensive checks
+	if m.Character.HealthMax.Value <= 0 {
```
Overkill - I'll leave as is for now. It just adds extra code without real benefit.

MorquinDevlar commented on internal/configs/config.userinterface.go:
```diff
@@ -0,0 +1,72 @@
+		f.TimeShort = `Jan 2 '06 3:04PM`
+	}
+}
+
```
Will ignore. Trivial nitpick.

MorquinDevlar committed:
Refactored ItemLookData to use string-based type comparisons

Changed:
  - Replaced boolean fields (IsWeapon, IsReadable, etc.) with Type and
Subtype strings
  - Replaced IsCursed/IsEnchanted booleans with extensible Adjectives
map
  - Updated templates to use string equality checks (e.g., {{if eq .Type
"readable"}})

Fixed:
  - Corrected TimeoutMods YAML formatting from multi-line to single-line

MorquinDevlar commented:
I am making a few more changes to the code, just to clean things up a little. Moving this to a draft while I do it :) Let me know if the migration is still something you feel we should do.

We can do it on the fly when we read the file, keeping compatibility with the old TextFormats but only when reading the config file, or we can make a mere elaborate system where we re-write the actual config file to support the new system permanently.

MorquinDevlar committed:
Removed deprecated TextFormats config structure

Removed:
  - TextFormats struct definition and field from main Config struct
  - TextFormats.Validate() call from config validation process
  - config.textformats.go file containing backwards compatibility
function
  - Duplicate configuration entries in server output

Changed:
  - All references from GetTextFormatsConfig() to
GetUserInterfaceConfig().Formats
  - Direct TextFormats field access to UserInterface.Formats field
access
  - Config structure to use single source of truth for format settings

Fixed:
  - Duplicate prompt settings appearing in server set command output
  - Confusion from having both TextFormats.Prompt and
UserInterface.Formats.Prompt
  - Inconsistent config access patterns across the codebase

Updated files:
  - internal/configs/configs.go - removed TextFormats field and
validation
  - internal/mobcommands/go.go - updated to use
GetUserInterfaceConfig().Formats
  - internal/usercommands/go.go - updated to use
GetUserInterfaceConfig().Formats
  - internal/usercommands/history.go - updated to use
GetUserInterfaceConfig().Formats
  - internal/usercommands/set.go - updated to use
GetUserInterfaceConfig().Formats
  - internal/usercommands/start.go - updated to use
GetUserInterfaceConfig().Formats
  - internal/users/inbox.go - updated direct access to
UserInterface.Formats
  - internal/users/userrecord.prompt.go - updated to use
GetUserInterfaceConfig().Formats
  - modules/gmcp/gmcp.Game.go - updated direct access to
UserInterface.Formats
